core/src/main/java/org/bf2/srs/fleetmanager/execution/impl/workers/provision/CreateSubscriptionWorker.java

```diff
@@ -0,0 +1,174 @@
+
+            // Try to consume some quota from AMS for the appropriate resource type (standard or eval).  If successful
+            // we'll get back a subscriptionId - if not we'll throw an exception.
+
+            try {
+                String subscriptionId = accountManagementService.createResource(accountInfo, resourceType);
+                registry.setSubscriptionId(subscriptionId);
+            } catch (ResourceLimitReachedException rlre) {
+                log.warn("Resource limit reached:", rlre);
+                failedProvisioning = true;
+                provisioningError = rlre.getMessage();
+            }
+
+            // Convert to registry data
+            RegistryInstanceTypeValueDto instanceType = resourceTypeToInstanceType(resourceType);
+            registry.setInstanceType(instanceType.value());
+        }
+
+        // NOTE: Failure point 2
+        if (failedProvisioning) {
+            registry.setStatus(RegistryStatusValueDto.FAILED.value());
+            registry.setFailedReason(provisioningError);
+            storage.createOrUpdateRegistry(registry);
+        } else {
+            //We only try to schedule the instance if the subscription has been successfully created.
+            registry.setStatus(RegistryStatusValueDto.ACCEPTED.value());
+            storage.createOrUpdateRegistry(registry);
+            ctl.delay(() -> tasks.submit(ScheduleRegistryTask.builder().registryId(registry.getId()).build()));
+        }
+    }
+
+    private ResourceType determineResourceType(AccountInfo accountInfo) throws AccountManagementServiceException, EvalInstancesNotAllowedException, TooManyEvalInstancesForUserException {
+        final ResourceType resourceType = evalInstancesOnlyEnabled ?
+                ResourceType.REGISTRY_INSTANCE_EVAL : accountManagementService.determineAllowedResourceType(accountInfo);
+
+        if (resourceType == ResourceType.REGISTRY_INSTANCE_EVAL) {
+            checkEvalInstancesPermission(accountInfo);
+        }
+
+        return resourceType;
+    }
+
+    @Override
+    @Transactional
+    public void finallyExecute(Task aTask, WorkerContext ctl, Optional<Exception> error) throws Exception {
+        CreateSubscriptionTask task = (CreateSubscriptionTask) aTask;
+
+        // SUCCESS STATE
+        Optional<RegistryData> registryOpt = storage.getRegistryById(task.getRegistryId());
+        if (registryOpt.isPresent() && registryOpt.get().getInstanceType() != null)
+            return;
+
+        if (registryOpt.isPresent()) {
+            if (registryOpt.get().getSubscriptionId() != null) {
+                accountManagementService.deleteSubscription(registryOpt.get().getSubscriptionId());
+                log.warn("Returned subscription {} since something wen wrong during provisioning phase:", registryOpt.get().getSubscriptionId());
+            }
+            log.warn("Deleting registry data with registry {}", registryOpt.get());
+            storage.deleteRegistry(task.getRegistryId());
```

Comment:
I think we should keep it in the failed state instead of deleting, so the error/failed reason does not disappear.